---
title: "Selecting model cell lines for studying the phosphorylation signature of signaling pathways"
output: github_document
---

This is the modified version of the original analysis. We applied for an R01 for studying the signature of phosphorylation pathway using the Kinobeads (Golkowski et al. 2017. JPR). We hope to find the specific phosphorylation pattern for a given signal pathway. For example, based on the RNA-seq data from CCLE and TCGA, FGFR and EGFR are often co-expressed, and in fact, a lot of the phosphorylation events are common between these two pathways. So can we deconvolute them using Kinobeads? 

In the first version of this R01 application, we were criticized by choosing the model cell lines arbitrarily. To solve this, I analyzed the CCLE RNA expression data and chose cell lines based on their expression level of the pathway components. The pathway information was drawn from the reactome.org. The script can be found in https://github.com/ht-lau/cells_for_pathway_analysis

Fortunately, we are awarded the R01 after the revision (in a separated application cycle). To start the project, I re-visited the analysis and found that, using all of the pathway components as a guide can be a bad idea. For example, It is obvious that some of the cell lines can express the EGFR pathway components at a high level but do not express EGFR. This is also true for the other pathways. So I now think that using the expression of the essential component(s), e.g. the receptors, will be more appropriate.  

Followings are the signaling pathways chosen. Description is the path to find them in reactome.org

EGF: Signal Transduction, Signaling by RTKs, Signaling by EGFR
FGF: Signal Transduction, Signaling by RTKs, Signaling by FGFR
HGF: Signal Transduction, Signaling by RTKs, Signaling by MET
TGFb: Signal Transduction, Signaling by TGF-beta, Signaling by TGF-beta receptor
IL6: Immune Sys, Cytokine Signaling, Signaling by IL, IL-6
WNTs: Signal Transduction, Signaling by WNT, TCF dependent signaling in response to WNT
SHH: Signal Transduction, Signaling by Hedgehog
Estrogen: Signal Transduction, Signaling by NR, ESR-mediated signaling
Glucose: Signal Transduction, Signaling by RTK, Signaling by Insulin receptor
Oxygen: Cell. responses to ext. stimuli, Cell. response to stress, hypoxia
PPMs/ECM: Extracellular matrix organization, Integrin cell surface interactions

```{r packages, message=FALSE, echo=FALSE}
library(tidyverse)
library(ggthemes)
library(ggpubr)
library(readxl)
library(writexl)
library(qdapTools)  # required to turn the gene.list into df, list2df()
library(VennDiagram)
library(colorRamps)
library(RColorBrewer)
library(gplots)
```

```{r reactome gene list + TCGA oncogenes, echo=FALSE}

# This part deal with the data I downloaded from the Reactome.
# Mainly, I want to extract the genes involved in each pathway.
# This will produce a list that I can use to match with the CCLE expression data
# Note that these information is not used in this version, but I think it maybe valuable to keep these,
# in case I will have to re-visit the data

# read the files containing the signaling components into a list
signaling.files <- list.files(path = "signaling_genes", pattern = "by_",
                              full.names = TRUE)            # full name extract the path

# use the file list created above to read the files into a list of df
# this dfs contain the data from reactome, which have to clean up further to extract the genes
gene.list <- lapply(signaling.files, read.delim, header = TRUE, stringsAsFactors = FALSE)

# extract the pathway names from the file list
pathway.names <- lapply(signaling.files, str_sub, 20, -5)

# add pathway names back to the file list
names(gene.list) <- pathway.names


# The gene.list contains a lot of information that I don't need from the Reactome.
# For example, metabolites involved in the pathway.
# So write a function to clean it up.

extract.genes <- function(x, MoleculeType, MoleculeName) {
  x <- x %>% 
    filter(MoleculeType == "Proteins") %>% 
    separate(MoleculeName, into = c("Uniprot", "Gene.names"), sep = " ")
  x <- x[!duplicated(x$Gene.names), ]
  x <- x$Gene.names
}

# clean up the gene list with the new function
gene.list <- lapply(gene.list, FUN = extract.genes)
names(gene.list) <- str_c("Reactome.", names(gene.list))

```

To visualize the number of genes in each pathway.

```{r plotting the pathways, echo=FALSE}

gene.list.long <- list2df(gene.list)
names(gene.list.long)[1:2] <- c("Gene.names", "Pathways")
gene.list.long <- gene.list.long %>% separate(Pathways, into = c("Sources", "Pathways"), sep = "\\.")

ggplot(gene.list.long) +
  geom_bar(mapping = aes(x = Pathways)) +
  theme(axis.text.x = element_text(angle = 40, hjust = 1),
        panel.background = element_rect(fill = "white", colour = "grey50")) +
  ylab("Gene counts") + xlab("Pathway")

```

Load the gene expression data downloaded from CCLE. Then extract the pathway component genes with the expression data. Results in 11 list of pathway gene expression. 

```{r match the signaling component and calculate the sum of expressions, echo=FALSE}

# import the previously produced normalized CCLE expression data. DEseq2 normalized RNA-seq read counts
nor.ccle <- read.delim("Normalized_CCLE_RNAseq.txt", header = TRUE, stringsAsFactors = FALSE)
nor.ccle[1:5, 1:5]

# create a new list to whole the list of df with pathway component expression values
ccle.pathways <- list()

# match the whole CCLE expression data with the list containing pathway components
# then add Ensembl IDs as the gene names
for (i in 1:length(gene.list)) {
  ccle.pathways[[i]] <- nor.ccle[nor.ccle$Gene.names %in% gene.list[[i]],]
#  row.names(ccle.pathways[[i]]) <- ccle.pathways[[i]][,1]  # this line add the ensemble ID to row.names, but is not useful
}

# rename of the dfs in the list with each respective signaling pathway
names(ccle.pathways) <- names(gene.list)
ccle.pathways[[1]][1:5,1:5]

```

To select the cell line for studying a particular pathway, I will have to make sure that the essential pathway component(s) is expressed. For example, EGFR will have to be expressed in the cell chosen for studying the EGFR signaling; any of the FGFRs have to be expressed for the FGFR signaling model cells. My initial idea is to choose the cells with high essential gene expression. However, it may cause the pathway to be activated constitutively. To make sure the pathways are inducible thus can be controlled experimentally, I think a moderate/median expression will be more appropriate. 

Write a function to do the selection.

```{r cell selecting function}

# I made a table containing the essential genes in each pathway, e.g. EGFR in EGF signaling
essential <- read.delim("signaling_genes/essential_each_pathway.txt", stringsAsFactors = FALSE)
essential[1:5,]

# write a funtion to filter out cells that do not meet a threshold expression of the essential genes
# Input argument will be the essential gene data frame, from above
# The gene expression data (ccle.pathways) is already in the function 
# This function will,
#   1. filter the cells, if retain.percentile is provided. Cells that do not meet the criteria will be filled with 0
#   2. Export the expression of the essential gene(s). If retain.percentile is not provided.

filter.essential.genes <- function(essential.df,        # essential genes df
                                   retain.percentile = NULL    # a length-of-two vector with 
                                                               # the lower and upper expression to be kept
                                                               # NULL will output the expression of the essential genes
                                   ) { 
  pathway <- essential.df["Pathway"]
  essential.gene <- essential.df["Gene"]
  # the expression data is a list of data frame
  # So just extract the pathway data frame that match the pathway name
  filtered.pathway <- ccle.pathways[[pathway]]
  
  # make a new data frame holding the essential gene expression
  essential.gene.expression <- filtered.pathway[grepl(x = filtered.pathway$Gene.names,
                                                      pattern = essential.gene), ]
  
  
  if (length(retain.percentile) == 0) {
    return(essential.gene.expression)
  } else {
    # I think the cells should be expressing the essential genes, but not over-expressing. 
    # Like EGFR will auto-phosphorylate then it will be very hard to turn off the signal.
    # I will test, and remove the cells that are not within the expression range.
  
    # t(apply()) here because i want to fill two columns. 
    # If I don't t(), apply() will fill up percent.5 before percent.6,
    # meaning that percent.5 will have 50 %-tile and 60 %-tile of gene 1 in row 1 and row 2
    # then gene 2 in row 3 and row 4, etc
    essential.gene.expression[, c("low.bound", "hi.bound")] <- t(apply(essential.gene.expression[, 3:1078],
                                                                      1, quantile, probs = retain.percentile))
    # Test expression that are out of range
    test.threshold <- as.data.frame(apply(essential.gene.expression[, 3:1078, drop = FALSE], 2,
                                          function(x) x < essential.gene.expression$low.bound |
                                            x > essential.gene.expression$hi.bound))
  
    # The test was to see what genes are out of the target range,
    # so TRUE will be things that I don't want, i.e. extract cell lines to remove
  
    # The problem is that apply() turn the 1 row data frames into 1 column and n rows df...
    # So I have to make a if... else to test and fix it
    if (ncol(test.threshold) < 2) {
      unwant.cells <- test.threshold[test.threshold[,1] == "TRUE", , drop = FALSE]
      unwant.cells <- row.names(unwant.cells)
    } else {
#      unwant.cells <- names(test.threshold)[apply(test.threshold, 
#                                                  2, function(x) all(x == "TRUE"))]
      unwant.cells <- names(test.threshold)[apply(test.threshold, 
                                                  2, function(x) sum(x == "TRUE")/length(x) >= 0.6)]
    }
  
    # impute the expression data with a very low number for the cells that will be removed 
    filtered.pathway[, unwant.cells] <- 0
  
    return(filtered.pathway)
    }

}

```

Filter the cells. Cells that are not selected for a pathway will have gene expression of 0. 
```{r selecting cells}

# filtered cells will be for picking the cells to use. 
# Retain cells with essential gene expression between 0.4 and 0.6 percentile
filtered.reactomes <- apply(essential, 1, filter.essential.genes, retain.percentile = c(0.4, 0.6))
names(filtered.reactomes) <- sub(pattern = "Reactome.", replacement = "", x = essential$Pathway)

filtered.reactomes[[1]][1:6, 1:6]

```

Visualize the pathway component expression in the selected cells. And table the number of select cell in each pathway.
```{r plot components expression}

filtered.cells <- filtered.reactomes

filtered.cells <- lapply(filtered.cells, FUN = select, 3:1078)
filtered.cells.sumExp <- lapply(filtered.cells, function(x) data.frame(Sums = colSums(x)))

filtered.cells.sumExp <- Reduce(x = filtered.cells.sumExp, f = function(x,y) cbind(x,y))
names(filtered.cells.sumExp) <- names(filtered.reactomes)

filtered.cells <- filtered.cells.sumExp %>% 
  gather(key = "Pathways", value = "Sum.exp") %>%
  filter(Sum.exp != 0)

reactome.box <- ggplot(filtered.cells) +
  geom_boxplot(mapping = aes(x = Pathways, y = log10(Sum.exp))) +
  theme(axis.text.x = element_text(angle=45, hjust = 1),
        axis.title.x = element_blank(),
        panel.background = element_rect(colour = "black", fill = "white")) + 
  ggtitle("Sum expression of pathway components")

reactome.box

data.frame(table(filtered.cells$Pathways))

```

```{r export, echo=FALSE}

filtered.cells.sumExp$Non0 <- apply(filtered.cells.sumExp, 1, 
                                    function(x) sum(x != 0))

cells.with.six.pathways <- filtered.cells.sumExp %>% 
  mutate(Cells = rownames(filtered.cells.sumExp)) %>%
  filter(Non0 > 5) %>%
  select(-Non0) %>%
  gather(key = "Pathways", value = "Expression", -Cells) %>%
  spread(key = Cells, value = Expression) %>%
  rename(Gene.names = "Pathways") # I need to match the data to the sum data later, by the column names

#write.table(filtered.cells.sumExp, "SumExpressionAllCells.txt", quote = FALSE, append = FALSE, sep = "\t")

picked.cells <- rownames(filtered.cells.sumExp)[filtered.cells.sumExp$Non0 > 5]

picked.cells.expression <- lapply(ccle.pathways[1:11], select, Gene.names, picked.cells)

column.names <- names(cells.with.six.pathways)
for (i in 1:11) {
  picked.cells.expression[[i]][nrow(picked.cells.expression[[i]]) + 1, column.names] <- 
    cells.with.six.pathways[i, ]
}

cleanup <- function(x) {
  x <- data.frame(t(x))
  names(x) <- unname(unlist(x[1, ]))
  x <- x[-1,]
  colnames(x)[ncol(x)] <- "Can.be.used"
  x$Cells <- rownames(x)
  x <- x[, c(ncol(x), ncol(x)-1, 1:(ncol(x)-2))]
  x <- type.convert(x)
  x$Can.be.used <- ifelse(x$Can.be.used == 0, "", "+")
  return(x)
}
picked.cells.expression <- lapply(picked.cells.expression, cleanup)

#write_xlsx(picked.cells.expression, "ModelCells.xlsx")

```

Check the expression level of essential gene(s) in the selected cells in all of the CCLE cells. Box is the expression of the essential gene(s) in all cell lines. Each coloured dot is a selected cell line. Plotted is the sum of the essential gene(s) expression, e.g. FGFR pathway has FGFR1 - FGFR4, y-axis is the sum of all FGFR expression. 

```{r see where the cells are at the expression spectrum, warning=FALSE}

# extract the essential genes from the whole ccle data
pathway.genes.expression.list <- apply(essential, 1, filter.essential.genes)
pathway.names <- sub(pattern = "Reactome.", replacement = "", x = essential$Pathway)
num.of.genes <- sapply(pathway.genes.expression.list, nrow)
pathway.genes.expression <- Reduce(x = pathway.genes.expression.list, f = function(x,y) rbind(x,y))
names(pathway.genes.expression)[1] <- "Pathways"
pathway.genes.expression$Pathways <- rep(pathway.names, times = num.of.genes)

pathway.genes.expression <- pathway.genes.expression %>%
  gather(key = "Cells", value = "Expression", 3:1078)

# tidy up the selected cells for plotting
selected.cells.for.plotting <- lapply(picked.cells.expression, 
                                      function(x) filter(x, Can.be.used == "+") %>% select(1:2))
#selected.cells.for.plotting <- lapply(selected.cells.for.plotting, function(x) names(x)[3] <- "Gene.x")

num.of.cells <- sapply(selected.cells.for.plotting, nrow)
selected.cells.for.plotting <- Reduce(x = selected.cells.for.plotting, f = function(x,y) rbind(x,y))
selected.cells.for.plotting$Pathways <- rep(pathway.names, times = num.of.cells)
selected.cells.for.plotting <- selected.cells.for.plotting %>%
  select(-Can.be.used) %>%
  left_join(., pathway.genes.expression, by = c("Cells", "Pathways"))
  
# plot
p <- ggplot() +
  geom_boxplot(data = pathway.genes.expression, mapping = aes(x = Pathways, y = log10(Expression))) +
  geom_point(data = selected.cells.for.plotting, 
               mapping = aes(x = factor(Pathways), y = log10(Expression),
                             colour = Cells)) +
  xlab("Expression of pathway marker genes") +
  theme(axis.text.x = element_text(angle=45, hjust = 1),
        panel.background = element_rect(colour = "black", fill = "white"))

p + guides(colour = FALSE)

```

```{r get the sum of marker genes expression, echo=FALSE}
# Exported the above results. But it will probably be good to know the actual marker gene expression. So lets add it.

pathway.genes.expression.list <- pathway.genes.expression.list # this is from above
pathway.genes.sum <- lapply(pathway.genes.expression.list, 
                                       function(x) data.frame(apply(x[-c(1,2)], 2, sum)))
pathway.genes.sum <- Reduce(x = pathway.genes.sum, 
                                       f = function(x,y) cbind(x,y))
names(pathway.genes.sum) <- pathway.names
pathway.genes.sum$Cells <- rownames(pathway.genes.sum)
pathway.genes.sum <- select(pathway.genes.sum, Cells, 1:11)

# get the percentile of sum expression, then group them into different classes
get.percentile.group <- function(x) {
  break.points <- quantile(x, probs = c(0, .05, .25, .4, .6, .75, .95, 1))
  cut(x, 
      breaks = break.points, 
      labels = c(-3, -2, -1, 0, 1, 2, 3),
      include.lowest = TRUE)            # in help, the (a,b] notation means that a is not included, 
                                        # so for 0 - 5 % (0,0.05], lowest number is excluded, include.lowest = TRUE change that
}

# make a new df to hold the groups 
pathway.genes.group <- pathway.genes.sum
pathway.genes.group[, 2:12] <- apply(pathway.genes.sum[, 2:12], 2, get.percentile.group )
pathway.genes.group$Num.of.medianExp <- apply(pathway.genes.group[,2:12], 1, 
                                              function(x) sum(x == 0))

# prepare and make the boxplot
sum.for.plotting <- gather(pathway.genes.sum, key = "Pathways", value = "SumExp", 2:12)
picked.cells.by.group <- pathway.genes.group$Cells[pathway.genes.group$Num.of.medianExp > 5]

g <- ggplot() +
  geom_boxplot(sum.for.plotting, mapping = aes(x = Pathways, y = log10(SumExp))) +
  geom_point(sum.for.plotting[sum.for.plotting$Cells %in% picked.cells.by.group,],
             mapping = aes(x = Pathways, y = log10(SumExp), colour = Cells)) +
  theme(axis.text.x = element_text(angle = 30, hjust = 1),
        panel.background = element_rect(colour = "black", fill = "white"))

```

The export data contains the perentile of essential gene(s) expression, -3 = 0 - 5%, -2 = <=25%, -1 = <=40%, 0 = 40% - 60%, etc. So cells with 0 of a particular pathway should be chosen for that pathway.   

```{r, echo = FALSE}
pathway.genes.group[1:3,]
```

I am wonder if there is any correlation between pathways, say FGFRs expression is correlated to EGFR expression? But no...

```{r}

pathway.genes.sum.log <- apply(pathway.genes.sum[,2:12], 2, function(x) log10(x))
psych::pairs.panels(pathway.genes.sum.log)

```

```{r echo=FALSE}
# export
#percentile.grouping <- list(pathway.genes.sum, pathway.genes.group)
#names(percentile.grouping) <- c("Sum.marker.genes.expression", "Percentile.group")
#write_xlsx(percentile.grouping, "ModelCells_percentileGroups.xlsx")
```

